File: 038_LowestCommonAncestorOfABinaryTree.cpp
// Topic: Binary Tree / Recursion
// Lowest Common Ancestor of a Binary Tree: Find lowest shared ancestor of two nodes.
// Time Complexity: O(n)
// Space Complexity: O(h)

#include <bits/stdc++.h>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q)
{
    if (!root || root == p || root == q)
        return root;
    TreeNode* L = lowestCommonAncestor(root->left, p, q);
    TreeNode* R = lowestCommonAncestor(root->right, p, q);
    if (L && R)
        return root;
    return L ? L : R;
}

int main()
{
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(5);
    root->right = new TreeNode(1);
    root->left->left = new TreeNode(6);
    root->left->right = new TreeNode(2);
    root->left->right->left = new TreeNode(7);
    root->left->right->right = new TreeNode(4);
    root->right->left = new TreeNode(0);
    root->right->right = new TreeNode(8);
    cout << lowestCommonAncestor(root, root->left, root->right)->val << "\n";               // 3
    cout << lowestCommonAncestor(root, root->left, root->left->right->right)->val << "\n";  // 5
    return 0;
}

File: 038_LowestCommonAncestorOfABinaryTree.py
"""
Topic: Tree / Recursion
Lowest Common Ancestor of a Binary Tree: classic recursive approach
Time Complexity: O(n)
Space Complexity: O(h)
"""

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def lowest_common_ancestor(root: Optional[TreeNode], p: TreeNode, q: TreeNode) -> Optional[TreeNode]:
    if not root or root == p or root == q:
        return root
    left = lowest_common_ancestor(root.left, p, q)
    right = lowest_common_ancestor(root.right, p, q)
    if left and right:
        return root
    return left if left else right

if __name__ == '__main__':
    # build simple tree
    a = TreeNode(3)
    a.left = TreeNode(5, TreeNode(6), TreeNode(2, TreeNode(7), TreeNode(4)))
    a.right = TreeNode(1, TreeNode(0), TreeNode(8))
    p = a.left
    q = a.right
    print(lowest_common_ancestor(a, p, q).val)  # 3
    print(lowest_common_ancestor(a, a.left.left, a.left.right.right).val)  # 5
